pkg/memtier/orderedMap.go

```diff
@@ -0,0 +1,113 @@
+package memtier
+
+import (
+	rbt "github.com/emirpasic/gods/trees/redblacktree"
+)
+
+// OrderedMap represents an ordered map data structure.
+// It combines the functionalities of a Red-Black Tree and a key-value map.
+// The keys are uint64 values, and the values are of any type specified by the user.
+type OrderedMap[V any] struct {
+	rbt.Tree // Tree is the underlying Red-Black Tree used for maintaining the order of keys.
+}
+
+func compareUint64(a, b interface{}) int {
+	if a.(uint64) < b.(uint64) {
+		return -1
+	} else if a.(uint64) > b.(uint64) {
+		return 1
+	}
+	return 0
+}
+
+// NewOrderedMap creates and returns a new instance of OrderedMap with the specified value type.
+func NewOrderedMap[V any]() *OrderedMap[V] {
+	return &OrderedMap[V]{
+		Tree: rbt.Tree{
+			// Set the Comparator function to compareUint64 for sorting keys.
+			Comparator: compareUint64,
+		},
+	}
+}
+
+// Get returns the value for a key.
+// If the key does not exist, returns nil and false.
+func (m *OrderedMap[V]) Get(key uint64) (value V, ok bool) {
+	v, ok := m.Ceiling(key)
+	if ok {
+		value = v.Value.(V)
+	}
+	return value, ok
+}
+
+// GetElement returns the value for a key.
+// If the key does not exist, returns nil.
+func (m *OrderedMap[V]) GetElement(key uint64) (value V) {
+	v, ok := m.Ceiling(key)
+	if ok {
+		value = v.Value.(V)
+	}
+	return value
+}
+
+// Set will set (or replace) a value for a key.
+// If the key already exists, it updates the value and returns false.
+// If the key is new, it adds the key-value pair and returns true.
+func (m *OrderedMap[V]) Set(key uint64, value V) bool {
+	m.Put(key, value)
+	return true
+}
+
+// Len returns the number of elements in the OrderedMap.
+func (m *OrderedMap[V]) Len() int {
+	return m.Size()
+}
+
+// FindRange returns a slice of values within the range of [start, end] in the OrderedMap.
+// It includes values whose keys are greater than or equal to start and less than or equal to end.
+func (m *OrderedMap[V]) FindRange(start, end uint64) (collections []V) {
+	// Find the nearest nodes for the start and end keys.
+	s, ok := m.Floor(start)
+	if !ok {
+		s = m.Left()
+	}
+	e, ok := m.Floor(end)
+	if !ok {
+		e = m.Right()
+	}
+
+	// If any of the nodes are nil or the keys are not within the range, return an empty collection.
+	if s == nil || e == nil {
```

Comment:
When an OrderedMap contains keys (addresses) 4, 6, 8 and we call FindRange(2,5), does the function return with empty list as because of `s` being nil?

Reply:
From L72, when s is nil (not found), s will be m.Left(), then [4] will be returned.